java/arcs/core/testutil/Fuzzing.kt

```diff
@@ -0,0 +1,344 @@
+ * a coupling between a specific pair of generators.
+ *
+ * The Function class (see below) allows the provision of explicit values when a T is expected.
+ */
+// TODO(shanestephens): Experiment with using A<Pair<I, O>> as input instead, and a generic
+// constructor for A<Pair<I, o>> given A<I> and T<I, O>.
+abstract class T<I, O> {
+  abstract operator fun invoke(i: I): O
+  fun asA(i: A<I>): A<O> {
+    class TAsA : A<O> {
+      override operator fun invoke(): O = this@T.invoke(i())
+    }
+    return TAsA()
+  }
+}
+
+/**
+ * A source of randomness.
+ */
+interface Seed {
+  fun nextDouble(): Double
+  fun nextLessThan(max: Int): Int
+  fun nextInRange(min: Int, max: Int): Int
+  fun nextInt(): Int
+}
+
+/**
+ * A seeded source of randomness. Will always produce the same sequence of values if given the
+ * same seed.
+ */
+open class SeededRandom(val seed: Int) : Seed {
+  val random = Random(seed)
+  override fun nextDouble(): Double = random.nextDouble()
+  override fun nextLessThan(max: Int): Int = random.nextInt(0, max)
+  override fun nextInRange(min: Int, max: Int): Int = random.nextInt(min, max + 1)
+  override fun nextInt(): Int = random.nextInt()
+  fun printSeed() {
+    println("Seed was $seed")
+  }
+}
+
+/**
+ * A source of randomness that is seeded by the current date & time.
+ */
+class DateSeededRandom : SeededRandom(LocalDateTime.now().hashCode())
+
+/**
+ * A generator (implementation of [A<T>]) that always produces the same, specified value.
+ */
+class Value<T>(val value: T) : A<T> {
+  override operator fun invoke(): T = value
+}
+
+/**
+ * A generator (implementation of [A<T>]) that produces each value in the provided list,
+ * in order. This generator restarts from the beginning of the list when all values are
+ * exhausted.
+ */
+class SequenceOf<T>(val values: List<T>) : A<T> {
+  var idx = 0
+  override operator fun invoke(): T {
+    val result = values[idx++]
+    if (idx == values.size) {
+      idx = 0
+    }
+    return result
+  }
+}
+
+/**
+ * A generator (implementation of [A<T>]) that randomly chooses from a provided list each time
+ * it's invoked.
+ */
+class ChooseFromList<T>(val s: Seed, val values: List<T>) : A<T> {
+  override operator fun invoke(): T {
+    return values[s.nextLessThan(values.size)]
+  }
+}
+
+/**
+ * A generator (implementation of [A<List<T>>]) that randomly produces a list of the provided
+ * length, using the provided generator).
+ */
+class ListOf<T>(val generator: A<T>, val length: A<Int>) : A<List<T>> {
+  override operator fun invoke(): List<T> {
+    return (1..length()).map { generator() }
+  }
+}
+
+/**
+ * A generator (implementation of [A<Map<T, U>>]) that randomly produces a map with the provided
+ * number of entries, with keys drawn from the provided key generator and values drawn from
+ * the provided value generator.
+ */
+class MapOf<T, U>(val key: A<T>, val value: A<U>, val entries: A<Int>) : A<Map<T, U>> {
```

Comment:
entries should be just an int, why does it need to be A<int>?

Reply:
if you want to generate a map of random size, you need to provide a random variable that controls the size. This is easily specialized to a map of a specific size (use Value), but going in the other direction (if this just took a value and you wanted a random size) is harder & more annoying.